File: resolver-templateruntime/src/main/java/com/iflytek/cyber/resolver/templateruntime/fragment/player/PlayerInfoFragment.java
package com.iflytek.cyber.resolver.templateruntime.fragment.player;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.VectorDrawable;
import android.media.session.PlaybackState;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.CircularProgressDrawable;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.iflytek.cyber.resolver.audioplayer.AudioPlayerResolver;
import com.iflytek.cyber.resolver.audioplayer.service.AudioPlayer;
import com.iflytek.cyber.resolver.audioplayer.service.AudioPlayerService;
import com.iflytek.cyber.resolver.audioplayer.service.model.AudioItem;
import com.iflytek.cyber.resolver.playbackcontroller.PlaybackControllerResolver;
import com.iflytek.cyber.resolver.templateruntime.GlideApp;
import com.iflytek.cyber.resolver.templateruntime.R;

import java.util.Locale;

public class PlayerInfoFragment extends Fragment implements AudioPlayer.PlayerInfoCallback, View.OnClickListener, SeekBar.OnSeekBarChangeListener {
    private static final String TAG_PREVIOUS = "PREVIOUS";
    private static final String TAG_NEXT = "NEXT";
    private static final String TAG_PLAY_PAUSE = "PLAY_PAUSE";

    private OnAlbumCallback onAlbumCallback;

    private AudioPlayerResolver audioPlayerResolver;
    private PlaybackControllerResolver playbackControllerResolver;

    private TextView tvTitle;
    private TextView tvArtist;
    private TextView tvSlogan;
    private ImageView ivPrevious;
    private ImageView ivPlayPause;
    private ImageView ivNext;
    private ImageView ivAlbum;
    private TextView tvCurrentPosition;
    private TextView tvDuration;
    private ImageView ivProviderLogo;
    private TextView tvProviderName;

    private SeekBar seekBar;
    private boolean seekBarDragging = false;
    private int seekBarProgressTarget = -1;

    private AudioItem currentAudioItem;
    private DataBindingCallback callback;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_player_info, container, false);
        tvTitle = view.findViewById(R.id.title);
        tvArtist = view.findViewById(R.id.artist);
        seekBar = view.findViewById(R.id.seek_bar);
        tvSlogan = view.findViewById(R.id.slogan);
        ivPlayPause = view.findViewById(R.id.play_pause);
        ivNext = view.findViewById(R.id.skip_next);
        ivPrevious = view.findViewById(R.id.skip_previous);
        tvCurrentPosition = view.findViewById(R.id.current_position);
        tvDuration = view.findViewById(R.id.duration);
        ivProviderLogo = view.findViewById(R.id.music_provider_logo);
        tvProviderName = view.findViewById(R.id.music_provider_name);
        ivAlbum = view.findViewById(R.id.album);
        return view;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        checker.start();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ivPrevious.setOnClickListener(this);
        ivPrevious.setTag(TAG_PREVIOUS);
        ivNext.setOnClickListener(this);
        ivNext.setTag(TAG_NEXT);
        ivPlayPause.setOnClickListener(this);
        ivPlayPause.setTag(TAG_PLAY_PAUSE);

        seekBar.setMax(100);
        seekBar.setEnabled(false);
        seekBar.setOnSeekBarChangeListener(this);

        ivAlbum.setImageResource(R.drawable.ic_default_album);
        ivAlbum.post(new Runnable() {
            @Override
            public void run() {
                if (onAlbumCallback != null) {
                    Bitmap bitmap = getBitmapFromVectorDrawable(R.drawable.ic_default_album);
                    onAlbumCallback.onAlbumChanged(bitmap);
                }
            }
        });

        AudioPlayer audioPlayer = getAudioPlayer();
        if (audioPlayer != null)
            audioPlayer.setPlayerInfoCallback(this);

    }

    public void setup(AudioPlayerResolver audioPlayerResolver) {
        this.audioPlayerResolver = audioPlayerResolver;
    }

    public void setup(PlaybackControllerResolver playbackControllerResolver) {
        this.playbackControllerResolver = playbackControllerResolver;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    public OnAlbumCallback getOnAlbumCallback() {
        return onAlbumCallback;
    }

    public void setOnAlbumCallback(OnAlbumCallback onAlbumCallback) {
        this.onAlbumCallback = onAlbumCallback;
    }

    public AudioItem getCurrentAudioItem() {
        return currentAudioItem;
    }

    @Override
    public void onAudioItemUpdated(final AudioItem audioItem) {
        currentAudioItem = audioItem;
        if (audioItem == null) {
            if (callback != null)
                callback.onDataBind(null, null);
            tvTitle.setText(null);
            tvArtist.setText(null);
            tvArtist.setVisibility(View.GONE);
            tvTitle.setVisibility(View.GONE);
            tvSlogan.setVisibility(View.VISIBLE);
            tvCurrentPosition.setText(format(0));
            tvDuration.setText(format(0));
            ivAlbum.setImageResource(R.drawable.ic_default_album);
            ivAlbum.post(new Runnable() {
                @Override
                public void run() {
                    if (onAlbumCallback != null) {
                        Bitmap bitmap = getBitmapFromVectorDrawable(R.drawable.ic_default_album);
                        onAlbumCallback.onAlbumChanged(bitmap);
                    }
                }
            });
            ivPlayPause.setImageResource(R.drawable.ic_play_circle_outline_white_24dp);
        } else if (!TextUtils.isEmpty(audioItem.audioItemId)) {
            PlayerInfoPayload payload = PlayerInfoMap.get().get(audioItem.audioItemId);
            if (payload != null) {
                if (callback != null)
                    callback.onDataBind(audioItem, payload);
                tvTitle.setText(payload.content.title);
                tvArtist.setText(payload.content.titleSubtext1);
                tvSlogan.setVisibility(View.GONE);
                tvArtist.setVisibility(View.VISIBLE);
                tvTitle.setVisibility(View.VISIBLE);
                if (payload.content.art != null && payload.content.art.sources != null
                        && payload.content.art.sources.size() > 0) {
                    GlideApp.with(getContext())
                            .asBitmap()
                            .load(payload.content.art.sources.get(0).url)
                            .placeholder(R.drawable.ic_default_album)
                            .error(R.drawable.ic_default_album)
                            .listener(new RequestListener<Bitmap>() {
                                @Override
                                public boolean onLoadFailed(@Nullable GlideException e,
                                                            Object model,
                                                            Target<Bitmap> target,
                                                            boolean isFirstResource) {
                                    return false;
                                }

                                @Override
                                public boolean onResourceReady(Bitmap resource,
                                                               Object model,
                                                               Target<Bitmap> target,
                                                               DataSource dataSource,
                                                               boolean isFirstResource) {
                                    if (onAlbumCallback != null)
                                        onAlbumCallback.onAlbumChanged(resource);
                                    return false;
                                }
                            })
                            .into(ivAlbum);
                } else {
                    ivAlbum.setImageResource(R.drawable.ic_default_album);
                    ivAlbum.post(new Runnable() {
                        @Override
                        public void run() {
                            if (onAlbumCallback != null) {
                                Bitmap bitmap = getBitmapFromVectorDrawable(R.drawable.ic_default_album);
                                onAlbumCallback.onAlbumChanged(bitmap);
                            }
                        }
                    });
                }
                if (payload.content.provider != null &&
                        payload.content.provider.logo != null &&
                        payload.content.provider.logo.sources.size() > 0) {
                    tvProviderName.setVisibility(View.GONE);
                    ivProviderLogo.setVisibility(View.VISIBLE);
                    Glide.with(getContext())
                            .load(payload.content.provider.logo.sources.get(0).url)
                            .listener(new RequestListener<Drawable>() {
                                @Override
                                public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                                    ivProviderLogo.setVisibility(View.GONE);
                                    tvProviderName.setVisibility(View.VISIBLE);
                                    return false;
                                }

                                @Override
                                public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                                    return false;
                                }
                            })
                            .into(ivProviderLogo);
                    tvProviderName.setText(getString(R.string.music_from_source, payload.content.provider.name));
                } else if (payload.content.provider != null && !TextUtils.isEmpty(payload.content.provider.name)) {
                    ivProviderLogo.setVisibility(View.GONE);
                    tvProviderName.setVisibility(View.VISIBLE);
                    tvProviderName.setText(getString(R.string.music_from_source, payload.content.provider.name));
                } else {
                    ivProviderLogo.setVisibility(View.GONE);
                    tvProviderName.setVisibility(View.GONE);
                }
            }
        }
    }

    public void requestFocus() {
        if (tvSlogan.getVisibility() == View.VISIBLE) {
            tvSlogan.requestFocus();
        }
        if (tvTitle.getVisibility() == View.VISIBLE) {
            tvTitle.requestFocus();
        }
    }


    public void abandonFocus() {
        tvSlogan.clearFocus();
        tvTitle.clearFocus();
    }

    private AudioPlayer getAudioPlayer() {
        if (audioPlayerResolver != null) {
            AudioPlayerService service = audioPlayerResolver.getPlayerService();
            if (service != null) {
                AudioPlayer audioPlayer = service.getAudioPlayer();
                if (audioPlayer != null) {

                } else {
                    Log.w(getClass().getSimpleName(), "AudioPlayer is null");
                }
                return audioPlayer;
            } else {
                Log.w(getClass().getSimpleName(), "AudioPlayerService is null");
            }
        } else
            Log.w(getClass().getSimpleName(), "AudioPlayerResolver is null");
        return null;
    }

    public void setDataBindingCallback(DataBindingCallback callback) {
        this.callback = callback;
    }

    @Override
    public void onProgressUpdated(int position) {
        AudioPlayer audioPlayer = getAudioPlayer();
        if (seekBarDragging)
            return;
        if (audioPlayer != null) {
            if (seekBar.getMax() != audioPlayer.getDuration())
                seekBar.setMax(audioPlayer.getDuration());
            seekBar.setProgress(position);
        } else {
            seekBar.setProgress(0);
        }
    }

    @Override
    public void onPlaybackStateUpdated(PlaybackState playbackState) {
        if (playbackState.getState() == PlaybackState.STATE_PLAYING) {
            ivPlayPause.setImageResource(R.drawable.ic_pause_circle_outline_white_24dp);
            seekBar.setEnabled(true);
        } else {
            ivPlayPause.setImageResource(R.drawable.ic_play_circle_outline_white_24dp);
        }
        if (playbackState.getState() == PlaybackState.STATE_STOPPED) {
            seekBar.setEnabled(false);
            seekBar.setProgress(0);
        }
    }

    @Override
    public void onClick(View v) {
        AudioPlayer audioPlayer = getAudioPlayer();
        if (audioPlayer == null)
            return;
        switch (v.getTag().toString()) {
            case TAG_PREVIOUS:
                playbackControllerResolver.postCommand(PlaybackControllerResolver.PREVIOUS_COMMAND_ISSUED);
                break;
            case TAG_NEXT:
                playbackControllerResolver.postCommand(PlaybackControllerResolver.NEXT_COMMAND_ISSUED);
                break;
            case TAG_PLAY_PAUSE:
                if (audioPlayer.isPlaying()) {
                    playbackControllerResolver.postCommand(PlaybackControllerResolver.PAUSE_COMMAND_ISSUED);
                    audioPlayer.pause();
                } else {
                    playbackControllerResolver.postCommand(PlaybackControllerResolver.PLAY_COMMAND_ISSUED);
                    audioPlayer.start();
                }
                break;
        }
    }

    private String format(int duration) {
        return String.format(Locale.getDefault(), "%2d:%02d", duration / 1000 / 60, duration / 1000 % 60);
    }

    @Override
    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
        AudioPlayer audioPlayer = getAudioPlayer();
        if (audioPlayer != null) {
            tvCurrentPosition.setText(format(progress));
            tvDuration.setText(format(audioPlayer.getDuration()));
        } else {
            tvCurrentPosition.setText(format(0));
            tvDuration.setText(format(0));
        }
        if (fromUser) {
            seekBarProgressTarget = progress;
        }
    }

    @Override
    public void onStartTrackingTouch(SeekBar seekBar) {
        seekBarDragging = true;
    }

    @Override
    public void onStopTrackingTouch(SeekBar seekBar) {
        seekBarDragging = false;

        if (seekBarProgressTarget != -1) {
            AudioPlayer audioPlayer = getAudioPlayer();
            if (audioPlayer != null) {
                audioPlayer.seekTo(seekBarProgressTarget);
            }
            seekBarProgressTarget = -1;
        }
    }

    public interface OnAlbumCallback {
        void onAlbumChanged(Bitmap bitmap);
    }

    private ServiceChecker checker = new ServiceChecker(new ServiceChecker.Callback() {
        @Override
        public boolean onCheck() {
            return getAudioPlayer() != null;
        }

        @Override
        public void onFinish() {
            AudioPlayer audioPlayer = getAudioPlayer();
            if (audioPlayer != null)
                audioPlayer.setPlayerInfoCallback(PlayerInfoFragment.this);
        }
    });

    private static class ServiceChecker extends Handler {
        private Callback callback;

        ServiceChecker(Callback callback) {
            this.callback = callback;
        }

        public void start() {
            sendEmptyMessage(0);
        }

        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            if (!callback.onCheck()) {
                sendEmptyMessageDelayed(0, 100);
            } else {
                callback.onFinish();
            }
        }

        interface Callback {
            boolean onCheck();

            void onFinish();
        }
    }

    private Bitmap getBitmapFromVectorDrawable(int drawableId) {
        Drawable drawable = ContextCompat.getDrawable(getContext(), drawableId);

        Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(),
                drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);

        return bitmap;
    }

    public interface DataBindingCallback {
        void onDataBind(AudioItem audioItem, PlayerInfoPayload payload);
    }
}

File: app/src/main/java/com/iflytek/cyber/iot/show/core/fragment/PlayerInfoFragment.kt
/*
 * Copyright (C) 2019 iFLYTEK CO.,LTD.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.iflytek.cyber.iot.show.core.fragment

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver
import android.widget.ImageView
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import androidx.core.widget.TextViewCompat
import androidx.navigation.Navigation
import cn.iflyos.sdk.android.impl.common.iFLYOSPlayerHandler
import cn.iflyos.sdk.android.impl.mediaplayer.MediaPlayerHandler
import cn.iflyos.sdk.android.impl.mediaplayer.MediaSourceFactory
import cn.iflyos.sdk.android.impl.template.SimpleTemplateDispatcher
import cn.iflyos.sdk.android.impl.template.TemplateRuntimeHandler
import cn.iflyos.sdk.android.v3.iFLYOSManager
import cn.iflyos.sdk.android.v3.iface.MediaPlayer
import cn.iflyos.sdk.android.v3.iface.PlatformInterface
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.target.Target
import com.bumptech.glide.request.transition.Transition
import com.google.gson.Gson
import com.iflytek.cyber.iot.show.core.LauncherActivity
import com.iflytek.cyber.iot.show.core.R
import com.iflytek.cyber.iot.show.core.model.Content
import com.iflytek.cyber.iot.show.core.model.ContentStorage
import com.iflytek.cyber.iot.show.core.model.TemplateContent
import com.iflytek.cyber.iot.show.core.utils.RoundedCornersTransformation
import com.warkiz.widget.IndicatorSeekBar
import com.warkiz.widget.OnSeekChangeListener
import com.warkiz.widget.SeekParams
import jp.wasabeef.blurry.Blurry
import java.util.*

class PlayerInfoFragment : BaseFragment(), View.OnClickListener, OnSeekChangeListener,
        MediaPlayerHandler.OnMediaStateChangedListener {

    private var tvTitle: TextView? = null
    private var tvArtist: TextView? = null
    private var ivPrevious: ImageView? = null
    private var ivPlayPause: ImageView? = null
    private var ivNext: ImageView? = null
    private var ivAlbum: ImageView? = null
    private var tvCurrentPosition: TextView? = null
    private var tvDuration: TextView? = null
    private var ivProviderLogo: ImageView? = null
    private var ivBack: ImageView? = null
    private var ivBlurBackground: ImageView? = null
    private var tvTipsTitle: TextView? = null

    private var seekBar: IndicatorSeekBar? = null
    private var seekBarDragging = false
    private var seekBarProgressTarget = -1
    private var albumHeight: Int = 0
    private var currentSourceId: String = ""

//    private val progressHandler = Handler()

    private var playerHandler: MediaPlayerHandler? = null

    private val manager = iFLYOSManager.getInstance()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_player_info, container, false)
        return view
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val iFLYOSPlayerHandler = manager.getHandler("AudioMediaPlayer") as? iFLYOSPlayerHandler
        playerHandler = iFLYOSPlayerHandler?.mediaPlayer as? MediaPlayerHandler
        playerHandler?.addOnMediaStateChangedListener(this)

        val handler = manager.getHandler(PlatformInterface.SpecialHandler.TEMPLATETUNTIME.value())
        if (handler is TemplateRuntimeHandler) {
            handler.registerTemplateDispatchedListener(simpleTemplateDispatcher)
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        ivBack = view.findViewById(R.id.iv_back)
        tvTitle = view.findViewById(R.id.title)
        tvTipsTitle = view.findViewById(R.id.tv_tips_title)
        tvArtist = view.findViewById(R.id.artist)
        seekBar = view.findViewById(R.id.seek_bar)
        ivPlayPause = view.findViewById(R.id.play_pause)
        ivNext = view.findViewById(R.id.skip_next)
        ivPrevious = view.findViewById(R.id.skip_previous)
        tvCurrentPosition = view.findViewById(R.id.current_position)
        tvDuration = view.findViewById(R.id.duration)
        ivProviderLogo = view.findViewById(R.id.music_provider_logo)
        ivAlbum = view.findViewById(R.id.album)
        ivBlurBackground = view.findViewById(R.id.iv_blur_background)

        ivBack?.setOnClickListener { v -> Navigation.findNavController(v).navigateUp() }

        ivPrevious?.setOnClickListener(this)
        ivPrevious?.tag = TAG_PREVIOUS
        ivNext?.setOnClickListener(this)
        ivNext?.tag = TAG_NEXT
        ivPlayPause?.setOnClickListener(this)
        ivPlayPause?.tag = TAG_PLAY_PAUSE

        seekBar?.max = 100f
        seekBar?.isEnabled = false
        seekBar?.onSeekChangeListener = this

        setupView()
    }

    private fun setupView() {
        if (!isAdded && launcher != null) {
            return
        }

        Blurry.with(launcher)
                .sampling(4)
                .radius(75)
                .color(Color.parseColor("#66212121"))
                .from(getBitmapFromVectorDrawable(R.drawable.cover_default))
                .into(ivBlurBackground!!)

        val content = ContentStorage.get().currentContent
        if (content == null) {
            tvTitle?.visibility = View.GONE
            tvTipsTitle?.visibility = View.VISIBLE
            tvTipsTitle?.text = getString(R.string.default_music_title)
            tvArtist?.text = null
            tvArtist?.visibility = View.GONE
            tvCurrentPosition?.text = format(0)
            tvDuration?.text = format(0)
            setDefaultCover()
            ivPlayPause?.setImageResource(R.drawable.ic_play_circle_outline_white_24dp)
        } else {
            tvTipsTitle?.visibility = View.GONE
            tvTitle?.visibility = View.VISIBLE
            tvTitle?.text = content.title
            tvArtist?.text = content.titleSubtext1
            tvArtist?.visibility = View.VISIBLE
            tvTitle?.visibility = View.VISIBLE
            if (content.art != null && content.art?.sources != null
                    && content.art?.sources?.size ?: 0 > 0) {
                if (albumHeight > 0) {
                    realSetCover(content, true, albumHeight / 16)
                } else {
                    ivAlbum?.let {
                        it.viewTreeObserver?.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
                            override fun onGlobalLayout() {
                                realSetCover(content, true, it.height / 16)
                                it.viewTreeObserver?.removeOnGlobalLayoutListener(this)
                            }
                        })
                    }
                }
            } else {
                setDefaultCover()
            }
            if (content.provider != null) {
                if (content.provider?.logo != null && content.provider?.logo?.sources?.size ?: 0 > 0) {
                    ivProviderLogo?.visibility = View.VISIBLE
                    ivProviderLogo?.let {
                        Glide.with(launcher!!)
                                .load(content.provider?.logo?.sources?.get(0)?.url)
                                .listener(object : RequestListener<Drawable> {
                                    override fun onLoadFailed(e: GlideException?, model: Any,
                                                              target: Target<Drawable>, isFirstResource: Boolean): Boolean {
                                        it.visibility = View.GONE
                                        return false
                                    }

                                    override fun onResourceReady(resource: Drawable, model: Any,
                                                                 target: Target<Drawable>, dataSource: DataSource,
                                                                 isFirstResource: Boolean): Boolean {
                                        return false
                                    }
                                })
                                .into(it)
                    }
                }
            } else {
                ivProviderLogo?.visibility = View.GONE
            }
        }

        playerHandler?.let { playerHandler ->
            if (playerHandler.isPlaying()) {
                onMediaStateChanged("AudioMediaPlayer", "", MediaPlayer.MediaState.PLAYING)
            } else {
                onMediaStateChanged("AudioMediaPlayer", "", MediaPlayer.MediaState.STOPPED)
            }
            tvCurrentPosition?.text = format(playerHandler.getPosition())
            tvDuration?.text = format(playerHandler.duration)
            seekBar?.max = playerHandler.duration.toFloat()
            seekBar?.setProgress(playerHandler.getPosition().toFloat())
        }

        ivBack?.post {
            ivBack?.let { imageView ->
                val padding = imageView.height * 12 / 56
                imageView.setPadding(padding, padding, padding, padding)
                imageView.setImageResource(R.drawable.ic_previous_white_32dp)
            }

            view?.let { view ->
                val height = view.height

                tvTitle?.let { textView ->
                    val textSize = height * 48 / 600

                    textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize.toFloat())
                    TextViewCompat.setLineHeight(textView, textSize * 56 / 48)
                }

                tvArtist?.let { textView ->
                    val textSize = height * 24 / 600

                    textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize.toFloat())
                    TextViewCompat.setLineHeight(textView, textSize * 40 / 24)

                    val layoutParams = textView.layoutParams as? ConstraintLayout.LayoutParams
                    layoutParams?.let { lp ->
                        lp.topMargin = textSize / 2
                        textView.layoutParams = lp
                    }
                }

                val timeTextSize = height * 16 / 600
                tvCurrentPosition?.setTextSize(TypedValue.COMPLEX_UNIT_PX, timeTextSize.toFloat())
                tvDuration?.setTextSize(TypedValue.COMPLEX_UNIT_PX, timeTextSize.toFloat())

                view.findViewById<View>(R.id.info_area)?.let { infoArea ->
                    val horizontalPadding = ivAlbum?.left ?: 0
                    infoArea.setPadding(horizontalPadding, 0, horizontalPadding, 0)
                }

                tvTipsTitle?.let { textView ->
                    val textSize = height * 48 / 600

                    textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize.toFloat())
                    TextViewCompat.setLineHeight(textView, textSize * 56 / 48)
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()

        (activity as? LauncherActivity)?.run {
            setStatusType(LauncherActivity.StatusType.PLAYER)
        }
    }

    private fun setDefaultCover() {
        val radius = IntArray(1)
        if (albumHeight > 0) {
            radius[0] = albumHeight / 16
            realSetCover(null, false, radius[0])
        } else {
            ivAlbum?.viewTreeObserver?.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
                override fun onGlobalLayout() {
                    albumHeight = ivAlbum!!.height
                    radius[0] = albumHeight / 16
                    realSetCover(null, false, radius[0])
                    ivAlbum?.viewTreeObserver?.removeOnGlobalLayoutListener(this)
                }
            })
        }
    }

    @SuppressLint("CheckResult")
    private fun realSetCover(content: Content?, shouldBlur: Boolean, radius: Int) {
        if (launcher == null) {
            return
        }

        if (shouldBlur && content?.art?.sources != null &&
                content.art?.sources?.size ?: 0 > 0) {
            Glide.with(ivAlbum!!.context)
                    .asBitmap()
                    .load(content.art?.sources?.get(0)?.url)
                    .apply(RequestOptions
                            .placeholderOf(R.drawable.cover_default)
                            .transform(RoundedCornersTransformation(radius, 0))
                            .error(R.drawable.cover_default))
                    .into(object : SimpleTarget<Bitmap>() {
                        override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
                            ivAlbum?.setImageBitmap(resource)
                            ivBlurBackground?.let { blur ->
                                Blurry.with(launcher)
                                        .sampling(4)
                                        .color(Color.parseColor("#66212121"))
                                        .from(resource)
                                        .into(blur)
                            }
                        }

                        override fun onLoadFailed(errorDrawable: Drawable?) {
                            super.onLoadFailed(errorDrawable)
                            ivAlbum?.setImageResource(R.drawable.cover_default)
                        }
                    })
        } else {
            ivAlbum?.let {
                Glide.with(launcher!!)
                        .load(R.drawable.cover_default)
                        .apply(RequestOptions
                                .placeholderOf(R.drawable.cover_default)
                                .transform(RoundedCornersTransformation(radius, 0)))
                        .into(it)
            }
        }
    }

    fun requestFocus() {
        if (tvTitle != null && tvTitle?.visibility == View.VISIBLE) {
            tvTitle?.requestFocus()
        }
    }

    fun abandonFocus() {
        if (tvTitle != null)
            tvTitle?.clearFocus()
    }

    override fun onClick(v: View) {
        Log.d("PlayerInfo", "onClick: ${v.tag}")
        when (v.tag.toString()) {
            TAG_PREVIOUS -> manager.executeCommand(iFLYOSManager.Command.CMD_PLAY_PREVIOUS)
            TAG_NEXT -> manager.executeCommand(iFLYOSManager.Command.CMD_PLAY_NEXT)
            TAG_PLAY_PAUSE -> if (playerHandler?.isPlaying() == true) {
                manager.executeCommand(iFLYOSManager.Command.CMD_PLAY_PAUSE)
            } else {
                manager.executeCommand(iFLYOSManager.Command.CMD_PLAY_PLAY)
            }
        }
    }

    private fun format(duration: Long): String {
        return String.format(Locale.getDefault(), "%2d:%02d", duration / 1000 / 60, duration / 1000 % 60)
    }

    private fun setSeekBarEnabled(enabled: Boolean, resetProgress: Boolean = true) {
        seekBar?.isEnabled = enabled
        if (enabled) {
            tvDuration?.visibility = View.VISIBLE
            tvCurrentPosition?.visibility = View.VISIBLE
        } else {
            if (resetProgress) {
                seekBar?.setProgress(0f)

                tvCurrentPosition?.visibility = View.INVISIBLE
                tvDuration?.visibility = View.INVISIBLE
            }
        }
    }

    override fun onMediaStateChanged(playerName: String, sourceId: String, state: MediaPlayer.MediaState) {
        if (playerName != PLAYER_NAME)
            return

        if (currentSourceId == sourceId) {
            playerHandler?.let { playerHandler ->
                if (playerHandler.currentMediaType != MediaSourceFactory.MediaType.HLS) {
                    setSeekBarEnabled(true)
                } else {
                    setSeekBarEnabled(false)
                }
            }
        }

        if (state == MediaPlayer.MediaState.PLAYING) {
            ivPlayPause?.setImageResource(R.drawable.ic_pause_circle_outline_white_24dp)
            setSeekBarEnabled(true)
            ContentStorage.get().isMusicPlaying = true
//            progressHandler.post(progressCallback)
        } else if (state == MediaPlayer.MediaState.PAUSED) {
            ivPlayPause?.setImageResource(R.drawable.ic_play_circle_outline_white_24dp)
            ContentStorage.get().isMusicPlaying = false
//            progressHandler.removeCallbacks(progressCallback)
        }
        if (state == MediaPlayer.MediaState.STOPPED
                || state == MediaPlayer.MediaState.FINISH) {
            ivPlayPause?.setImageResource(R.drawable.ic_play_circle_outline_white_24dp)
            ContentStorage.get().isMusicPlaying = false
            setSeekBarEnabled(enabled = false, resetProgress = false)
        }

        currentSourceId = sourceId
    }

    override fun onMediaError(playerName: String, mediaError: MediaPlayer.MediaError, message: String) {
        Log.e("PlayerInfo", "play media error: $mediaError   message is: $message")
    }

    override fun onPositionUpdated(playerName: String, position: Long) {
        if (playerName != PLAYER_NAME)
            return
        onPositionUpdated(position)
    }

    private val simpleTemplateDispatcher = object : SimpleTemplateDispatcher() {
        override fun onPlayerInfoDispatched(template: String) {
            val templateContent = Gson().fromJson(template, TemplateContent::class.java)
            val prevTemplate = ContentStorage.get().template
            ContentStorage.get().saveContent(templateContent.content)
            if (prevTemplate == null || prevTemplate.audioItemId != templateContent.audioItemId) {
                setupView()
                ContentStorage.get().saveTemplate(templateContent)
            }
        }
    }

    override fun onSeeking(seekParams: SeekParams) {
        playerHandler?.let { playerHandler ->
            tvCurrentPosition?.text = format(seekParams.progress.toLong())
            tvDuration?.text = format(playerHandler.duration)
        } ?: run {
            tvCurrentPosition?.text = format(0)
            tvDuration?.text = format(0)
        }
        if (seekParams.fromUser) {
            seekBarProgressTarget = seekParams.progress
        }
    }

    override fun onStartTrackingTouch(seekBar: IndicatorSeekBar) {
        seekBarDragging = true
    }

    override fun onStopTrackingTouch(seekBar: IndicatorSeekBar) {
        seekBarDragging = false

        if (seekBarProgressTarget != -1) {
            playerHandler?.setPosition(seekBarProgressTarget.toLong())
            seekBarProgressTarget = -1
        }
    }

    private fun onPositionUpdated(position: Long) {
        if (seekBarDragging) {
            return
        }

        if (position > 0) {
            ivPlayPause?.setImageResource(R.drawable.ic_pause_circle_outline_white_24dp)
            setSeekBarEnabled(true)
        } else {
            ivPlayPause?.setImageResource(R.drawable.ic_play_circle_outline_white_24dp)
        }

        playerHandler?.let { playerHandler ->
            if (playerHandler.currentMediaType == MediaSourceFactory.MediaType.HLS) {
                setSeekBarEnabled(false)
            } else {
                setSeekBarEnabled(true)
                if (seekBar?.max != playerHandler.duration.toFloat())
                    seekBar?.max = playerHandler.duration.toInt().toFloat()
                seekBar?.setProgress(position.toInt().toFloat())
            }
        } ?: run {
            seekBar?.setProgress(0f)
        }
    }

//    private val progressCallback = object : Runnable {
//        override fun run() {
//            val position = playerHandler?.getPosition()
//            if (position != null && playerHandler?.isPlaying() == true) {
//                onPositionUpdated(position)
//            }
//            progressHandler.postDelayed(this, 1000)
//        }
//    }

    private fun getBitmapFromVectorDrawable(drawableId: Int): Bitmap {
        val drawable = ContextCompat.getDrawable(launcher!!, drawableId)

        val bitmap = Bitmap.createBitmap(drawable!!.intrinsicWidth,
                drawable.intrinsicHeight, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)

        return bitmap
    }

    override fun onDestroy() {
        super.onDestroy()
        playerHandler?.removeOnMediaStateChangedListener(this)
//        progressHandler.removeCallbacks(progressCallback)
        val handler = manager.getHandler(PlatformInterface.SpecialHandler.TEMPLATETUNTIME.value())
        if (handler is TemplateRuntimeHandler) {
            handler.unregisterDispatchedListener(simpleTemplateDispatcher)
        }
    }

    companion object {
        private const val PLAYER_NAME = "AudioMediaPlayer"

        private const val TAG_PREVIOUS = "PREVIOUS"
        private const val TAG_NEXT = "NEXT"
        private const val TAG_PLAY_PAUSE = "PLAY_PAUSE"
    }
}
